mergin/report.py

```diff
@@ -0,0 +1,237 @@
+                self.area = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.area = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.area = self.measure(self.new_geom)
+        else:
+            # regardless of geometry change count as 1
+            self.metric = "count"
+            self.count = 1
+
+    def measure(self, geom):
+        """ Return length or area of geometry based on type """
+        # calculate geom length/area only if QGIS API is available
+        try:
+            from qgis.core import QgsGeometry, QgsDistanceArea, QgsCoordinateReferenceSystem, QgsCoordinateTransformContext
+        except ImportError:
+            return -1
+
+        d = QgsDistanceArea()
+        d.setEllipsoid('WGS84')
+        crs = QgsCoordinateReferenceSystem()
+        crs.createFromString(self.crs)
+        d.setSourceCrs(crs, QgsCoordinateTransformContext())
+        g = QgsGeometry()
+        wkb_header_length = parse_gpkgb_header_size(geom)
+        wkb_geom = geom[wkb_header_length:]
+        g.fromWkb(wkb_geom)
+        if self.metric == "length":
+            return d.measureLength(g)
+        elif self.metric == "area":
+            return d.measureArea(g)
+        else:
+            return 1
+
+
+class ChangesetReport:
+    """ Report (aggregate) from geopackage changeset (diff file) """
+    def __init__(self, changeset_reader, schema):
+        self.cr = changeset_reader
+        self.schema = schema
+        self.entries = []
+        # let's iterate through reader and populate entries
+        for entry in self.cr:
+            schema_table = next((t for t in schema if t["table"] == entry.table.name), None)
+            # get geometry index in both gpkg schema and diffs values
+            geom_idx = next((index for (index, col) in enumerate(schema_table["columns"]) if col["type"] == "geometry"),
+                            None)
+            if not geom_idx:
+                continue
+
+            geom_col = schema_table["columns"][geom_idx]["geometry"]
+            report_entry = ChangesetReportEntry(entry, geom_idx, geom_col)
+            self.entries.append(report_entry)
+
+    def report(self):
+        """ Aggregate entries by table and operation type and calculate quantity """
+        records = []
+        tables = defaultdict(list)
+
+        for obj in self.entries:
+            tables[obj.table].append(obj)
+
+        for table, entries in tables.items():
+            items = groupby(entries, lambda i: (i.operation, i.metric))
+            for k, v in items:
+                quantity_type = "_".join(k)
+                values = list(v)
+                quantity = sum([getattr(entry, k[1]) for entry in values])
+                records.append({
+                    "table": table,
+                    "quantity_type": quantity_type,
+                    "quantity": quantity
+                })
```

Comment:
Sorry I am a bit lost what is going on in this for-loop... this does the aggregation of individual changeset entries of a single gpkg and version to a single line, is that correct?

Reply:
single gpkg, single table, single version and all same quantity type into one row - I will make a comment on that